python/cudf/cudf/core/column/column.py

```diff
@@ -1368,11 +1369,30 @@ def as_column(arbitrary, nan_as_null=None, dtype=None, length=None):
                 pa.array(np.asarray(arbitrary), from_pandas=True),
                 dtype=arbitrary.dtype,
             )
-        else:
+        elif arbitrary.dtype.kind in ("O", "U"):
             data = as_column(
                 pa.array(arbitrary, from_pandas=nan_as_null),
                 dtype=arbitrary.dtype,
             )
+        else:
+            print(arbitrary.dtype)
+            arb_dtype = check_upcast_unsupported_dtype(arbitrary.dtype)
+
+            if arb_dtype != arbitrary.dtype:
+                # This cast is exclusively needed for halffloat to
+                # float until the following issues are resolved:
+                # https://github.com/apache/arrow/issues/2691
+                # https://issues.apache.org/jira/browse/ARROW-7242
+                arbitrary = arbitrary.astype(arb_dtype)
```

Comment:
We should be able to handle all of these situations by doing `cupy.asarray(arbitrary).astype(arb_dtype)` to both transfer less data to the GPU, and have faster typecasting by doing it on the GPU.

Instead of doing this here, we should handle it in the `__cuda_array_interface__` case above and just return a `uint` typed cupy array here for example.

Reply:
Made the changes, but when there is a `float128`, cupy will error. Hence for all `float`/`f` kind dtypes we will hit `check_upcast_unsupported_dtype` and error.

Reply:
Yea we should absolutely just error on `float128`.